sparkdown-document-views: full-suite runs fail one test per run, a different one each time

# `sparkdown-document-views` full-suite runs fail one test per run, a different one each time

## Summary

Running the whole `packages/sparkdown-document-views` suite intermittently fails **exactly one** test — but a *different* test on different runs. Every one of them passes when run in isolation. So the suite is not trustworthy as a gate right now: a red run says nothing about whether the change under test broke anything.

## Observed

Three consecutive full runs of `npx vitest run` in `packages/sparkdown-document-views`:

| run | working tree | result |
| --- | --- | --- |
| 1 | clean | **60 passed**, 4 skipped |
| 2 | one unrelated CSS-in-JS change (`EDITOR_THEME.ts` z-index) | 59 passed, **1 failed** — `dialogue-squish.test.ts > emits a blank-line separator between consecutive dialogue blocks even when the source 'blank' line has trailing whitespace` |
| 3 | that change stashed — i.e. identical to run 1 | 59 passed, **1 failed** — `dual-after-single-incremental.test.ts > inserting the dual pair into an already-rendered single dialogue matches from-scratch` |

Run 3 is the important one: same tree as run 1, different outcome, and a different test than run 2. Re-running `dialogue-squish.test.ts` alone immediately after run 2 passed.

The failure in run 2 was:

```
AssertionError: preview should contain at least 2 character cues: expected 1 to be greater than or equal to 2
  ❯ assertBlankBetweenCharacters test/dialogue-squish.test.ts:61:9
```

i.e. the rendered preview came back with less content than expected — consistent with reading the DOM before the render settled, rather than with a wrong result.

## Why it matters

This suite is the only automated check on the script-editor rendering path. While it fails randomly, it trains everyone to ignore a red run, which is worse than having no gate.

It also cost real time during #278: a failing `dialogue-squish` looked like a regression from an unrelated z-index change, and ruling that out took a stash plus two 100-second full runs.

## Where to look

Both failing tests render into a shared jsdom document and assert on the resulting DOM. The suspicion is a render/settle race that only shows under full-suite load (each run takes ~100s, most of it `environment`), not an ordering dependency between specific files — the two failures are in unrelated files and neither fails alone.

Worth checking:

- whether the tests await the same "rendered" signal the component actually uses, or a fixed tick/timeout that stops being long enough when the machine is loaded
- whether module-level state (a shared editor/document, a registry) leaks between files
- whether running with `--sequence.shuffle` or a single fork makes it deterministic, which would point at cross-file leakage rather than timing

## Repro

```
cd packages/sparkdown-document-views
npx vitest run          # repeat a few times
```

Expect a different single failure on some runs; each failing test passes on its own.

--- a/packages/sparkdown-document-views/test/dual-after-single-incremental.test.ts
+++ b/packages/sparkdown-document-views/test/dual-after-single-incremental.test.ts
@@ -10,7 +10,7 @@
 // fixture as dual-after-single.test.ts and assert the resulting DOM
 // matches a from-scratch render.
 
-import { ensureSyntaxTree, language } from "@codemirror/language";
+import { language } from "@codemirror/language";
 import { EditorState } from "@codemirror/state";
 import { EditorView } from "@codemirror/view";
 import { describe, expect, it } from "vitest";
@@ -18,6 +18,7 @@ import {
   SCREENPLAY_LANGUAGE_SUPPORT,
   default as screenplayFormatting,
 } from "../src/modules/screenplay-preview/utils/screenplayFormatting";
+import { settleParse } from "./helpers/parseSettle";
 
 const FIXTURE =
   `BUNNY:\n` +
@@ -56,8 +57,7 @@ const mount = (source: string): EditorView => {
 };
 
 const contentHTML = (view: EditorView): string => {
-  ensureSyntaxTree(view.state, view.state.doc.length, 30_000);
-  view.dispatch({});
+  settleParse(view);
   const c = view.dom.querySelector(".cm-content");
   if (!c) return "";
   return c.outerHTML.replace(
@@ -67,29 +67,24 @@ const contentHTML = (view: EditorView): string => {
 };
 
 describe("dual dialogue incremental construction", () => {
-  it(
-    "character-by-character insertion produces the same DOM as a from-scratch render",
-    { timeout: 30_000 },
-    () => {
-      // Start with nothing, build the fixture one character at a time.
-      // ~600 single-char dispatches; passes in ~2.5s alone but crosses
-      // the default 5s timeout when other test files are running in
-      // parallel workers. Matches the pattern dual-real-incremental
-      // already uses for long-running incremental tests.
-      const incremental = mount("");
-      for (let i = 0; i < FIXTURE.length; i++) {
-        incremental.dispatch({
-          changes: { from: i, insert: FIXTURE[i]! },
-        });
-      }
-      const incrementalHTML = contentHTML(incremental);
+  it("character-by-character insertion produces the same DOM as a from-scratch render", () => {
+    // Start with nothing, build the fixture one character at a time.
+    // ~600 single-char dispatches; passes in a few seconds alone but crosses
+    // the default 5s timeout when other test files are running in parallel
+    // workers — hence the suite-wide `testTimeout` in vitest.config.ts.
+    const incremental = mount("");
+    for (let i = 0; i < FIXTURE.length; i++) {
+      incremental.dispatch({
+        changes: { from: i, insert: FIXTURE[i]! },
+      });
+    }
+    const incrementalHTML = contentHTML(incremental);
 
-      const fullRebuild = mount(FIXTURE);
-      const fullRebuildHTML = contentHTML(fullRebuild);
+    const fullRebuild = mount(FIXTURE);
+    const fullRebuildHTML = contentHTML(fullRebuild);
 
-      expect(incrementalHTML).toBe(fullRebuildHTML);
-    },
-  );
+    expect(incrementalHTML).toBe(fullRebuildHTML);
+  });
 
   it("inserting the dual pair into an already-rendered single dialogue matches from-scratch", () => {
     // The case the user likely actually hit: had the BUNNY single + the

--- a/packages/sparkdown-document-views/test/helpers/parseSettle.ts
+++ b/packages/sparkdown-document-views/test/helpers/parseSettle.ts
@@ -1,0 +1,72 @@
+// Make CodeMirror's syntax tree deterministic before a test reads it.
+//
+// Why this exists (the cause of #281's one-random-failure-per-run):
+//
+// `@codemirror/language` parses on a WALL-CLOCK budget, not to completion.
+// `LanguageState.init` gives the initial parse 20ms over the first 3000
+// characters and calls `takeTree()` — truncating the tree — if it runs out.
+// So the tree a freshly-created `EditorState`/`EditorView` exposes depends on
+// how loaded the machine was at that instant. Under a full suite run (~100s,
+// most of it jsdom setup) a fixture that parses completely when run alone
+// parses only partway, and any assertion about "what the preview shows" reads
+// a document that's missing its tail. That's the "expected 1 to be greater
+// than or equal to 2" class of failure: a second character cue that simply
+// hadn't been parsed yet.
+//
+// The two functions below force the parse to finish AND commit the finished
+// tree where the reader will actually see it. Committing is the subtle part:
+//
+//   `ensureSyntaxTree(state, len, timeout)` finishes the parse and RETURNS the
+//   complete tree, but it only mutates the mutable `ParseContext` hanging off
+//   the state field. `syntaxTree(state)` reads `field.tree`, a snapshot taken
+//   when the field was constructed — so it still hands back the truncated
+//   tree. Callers must use the RETURN VALUE (`completeTree`) or dispatch a
+//   transaction to refresh the field (`settleParse`, via `forceParsing`).
+
+import { ensureSyntaxTree, forceParsing } from "@codemirror/language";
+import type { EditorState } from "@codemirror/state";
+import type { EditorView } from "@codemirror/view";
+import type { Tree } from "@lezer/common";
+
+// Generous: these are whole-fixture parses, and the point is to never let the
+// wall clock decide the result. A fixture that genuinely needs 30s of parsing
+// is a bug worth failing on.
+export const PARSE_TIMEOUT_MS = 30_000;
+
+/**
+ * Parse `state`'s document to completion and return the complete tree.
+ *
+ * Use the returned tree — do NOT follow this with `syntaxTree(state)`, which
+ * still returns the truncated snapshot (see the note at the top of the file).
+ */
+export const completeTree = (state: EditorState): Tree => {
+  const tree = ensureSyntaxTree(state, state.doc.length, PARSE_TIMEOUT_MS);
+  if (!tree) {
+    throw new Error(
+      `Parse did not complete within ${PARSE_TIMEOUT_MS}ms for a ${state.doc.length}-character document. ` +
+        `Either the parser is stuck or no language is configured on this state.`,
+    );
+  }
+  return tree;
+};
+
+/**
+ * Parse `view`'s document to completion and re-run the decoration field
+ * against the complete tree, so the rendered DOM is safe to read.
+ *
+ * `forceParsing` dispatches an empty transaction when the tree changed, which
+ * refreshes the language state field; `replaceDecorations.update` then sees a
+ * new tree and recomputes every decoration from it.
+ */
+export const settleParse = (view: EditorView): void => {
+  const complete = forceParsing(
+    view,
+    view.state.doc.length,
+    PARSE_TIMEOUT_MS,
+  );
+  if (!complete) {
+    throw new Error(
+      `Parse did not complete within ${PARSE_TIMEOUT_MS}ms for a ${view.state.doc.length}-character document.`,
+    );
+  }
+};

--- a/packages/sparkdown-document-views/test/helpers/previewText.ts
+++ b/packages/sparkdown-document-views/test/helpers/previewText.ts
@@ -19,7 +19,7 @@
 //      line. If anything remains, look up the enclosing block kind in the
 //      tree at the line's starting position and emit `<kind> visible`.
 
-import { ensureSyntaxTree, language, syntaxTree } from "@codemirror/language";
+import { language } from "@codemirror/language";
 import { EditorState } from "@codemirror/state";
 import type { Range } from "@codemirror/state";
 import type { Decoration } from "@codemirror/view";
@@ -29,10 +29,9 @@ import {
   SCREENPLAY_LANGUAGE_SUPPORT,
   decorate,
 } from "../../src/modules/screenplay-preview/utils/screenplayFormatting";
+import { completeTree } from "./parseSettle";
 
 type HideInterval = { from: number; to: number };
-
-const PARSE_TIMEOUT_MS = 30_000;
 
 // Build state with language facet via language.of() (Language.extension
 // getter doesn't activate the facet correctly in this test environment).
@@ -41,9 +40,10 @@ const parseSource = (source: string): { state: EditorState; tree: Tree } => {
     doc: source,
     extensions: [language.of(SCREENPLAY_LANGUAGE_SUPPORT.language)],
   });
-  ensureSyntaxTree(state, source.length, PARSE_TIMEOUT_MS);
-  const tree = syntaxTree(state);
-  return { state, tree };
+  // `completeTree`, NOT `syntaxTree(state)` — the latter returns the
+  // 20ms-budgeted truncated snapshot taken when the state was created, which
+  // is what made this extractor's output depend on machine load (#281).
+  return { state, tree: completeTree(state) };
 };
 
 const collectHiddenRanges = (

--- a/packages/sparkdown-document-views/test/helpers/renderPreview.ts
+++ b/packages/sparkdown-document-views/test/helpers/renderPreview.ts
@@ -14,13 +14,14 @@
 // inspection ("does the line have display:none? are all its children
 // visibility:hidden width:0?") to infer visible/invisible.
 
-import { forceParsing, language } from "@codemirror/language";
+import { language } from "@codemirror/language";
 import { EditorState } from "@codemirror/state";
 import { EditorView } from "@codemirror/view";
 import {
   SCREENPLAY_LANGUAGE_SUPPORT,
   default as screenplayFormatting,
 } from "../../src/modules/screenplay-preview/utils/screenplayFormatting";
+import { settleParse } from "./parseSettle";
 
 export type RenderedLine = {
   index: number;
@@ -69,6 +70,11 @@ export const renderPreview = (source: string): RenderResult => {
       }),
       parent,
     });
+    // Finish the parse and recompute decorations against the complete tree
+    // before sampling the DOM. Without this the rendered output reflects
+    // whatever the 20ms initial-parse budget happened to cover on this
+    // machine, at this moment — the root cause of #281's random failures.
+    settleParse(view);
     const content = parent.querySelector(".cm-content");
     const contentHTML = content ? content.outerHTML : "";
     const lines: RenderedLine[] = [];
@@ -140,10 +146,10 @@ export const extractVisibleText = (source: string): string[] => {
     // only gets whatever partial tree the parser produces synchronously within
     // its time budget — when cold, that can stop mid-document, leaving trailing
     // nodes (e.g. a scene-closing `end`) un-decorated and so leaking as raw
-    // text. forceParsing finishes the parse AND updates the view, so the
+    // text. settleParse finishes the parse AND updates the view, so the
     // decoration field recomputes from the full tree — making the rendered DOM
     // deterministic regardless of parser warm-up state.
-    forceParsing(view, source.length, 30_000);
+    settleParse(view);
     const win = parent.ownerDocument!.defaultView as any;
     const visibleTextOf = (el: Element): string => {
       const cs = win.getComputedStyle(el);

--- a/packages/sparkdown-document-views/test/incremental-sync.test.ts
+++ b/packages/sparkdown-document-views/test/incremental-sync.test.ts
@@ -21,6 +21,7 @@ import {
   SCREENPLAY_LANGUAGE_SUPPORT,
   default as screenplayFormatting,
 } from "../src/modules/screenplay-preview/utils/screenplayFormatting";
+import { settleParse } from "./helpers/parseSettle";
 
 const mount = (source: string): EditorView => {
   const parent = document.createElement("div");
@@ -45,6 +46,9 @@ const mount = (source: string): EditorView => {
 // line decorations, not how many estimated pixels were reserved for the
 // virtual scroll.
 const contentHTML = (view: EditorView): string => {
+  // Both sides must be sampled from a COMPLETE parse, or the comparison is
+  // really "did these two parses get the same wall-clock slice" (#281).
+  settleParse(view);
   const c = view.dom.querySelector(".cm-content");
   if (!c) return "";
   return c.outerHTML.replace(

--- a/packages/sparkdown-document-views/test/indented-vs-empty-blank-parity.test.ts
+++ b/packages/sparkdown-document-views/test/indented-vs-empty-blank-parity.test.ts
@@ -17,6 +17,7 @@ import {
   SCREENPLAY_LANGUAGE_SUPPORT,
   default as screenplayFormatting,
 } from "../src/modules/screenplay-preview/utils/screenplayFormatting";
+import { settleParse } from "./helpers/parseSettle";
 
 // Indented blank line between dialogue and following action.
 const FIXTURE_INDENTED_BLANK =
@@ -43,6 +44,9 @@ const mountAndExtractCollapseLine = (
       }),
       parent,
     });
+    // The initial parse is wall-clock budgeted, so the `collapse` line we're
+    // looking for may not exist yet on a loaded machine (#281).
+    settleParse(view);
     const lineEls = Array.from(
       view.dom.querySelectorAll(".cm-line"),
     ) as HTMLElement[];

--- a/packages/sparkdown-document-views/test/parse-settle.test.ts
+++ b/packages/sparkdown-document-views/test/parse-settle.test.ts
@@ -1,0 +1,105 @@
+// Regression guard for #281: "full-suite runs fail one test per run, a
+// different one each time".
+//
+// Root cause: `@codemirror/language` budgets the initial parse by wall clock
+// (20ms, over at most the first 3000 characters) and truncates the tree if it
+// runs out. Every helper that read the tree straight after creating a state or
+// view was therefore reading a document whose tail hadn't been parsed yet —
+// how much tail depended on how busy the machine was. Alone: fine. Inside a
+// ~100s full-suite run: a random test each time saw a short document and
+// failed on a count or a missing decoration.
+//
+// These tests use a fixture deliberately LONGER than the 3000-character
+// initial-parse viewport, which makes the truncation unconditional rather than
+// load-dependent. So they fail deterministically if a helper ever goes back to
+// reading the un-settled tree — no flake-chasing required.
+
+import { language, syntaxTree } from "@codemirror/language";
+import { EditorState } from "@codemirror/state";
+import { EditorView } from "@codemirror/view";
+import { describe, expect, it } from "vitest";
+import { completeTree, settleParse } from "./helpers/parseSettle";
+import { extractPreviewText } from "./helpers/previewText";
+import {
+  SCREENPLAY_LANGUAGE_SUPPORT,
+  default as screenplayFormatting,
+} from "../src/modules/screenplay-preview/utils/screenplayFormatting";
+
+// One self-contained dialogue exchange, ~150 characters.
+const EXCHANGE = (n: number) =>
+  `RAFFLES:\n` +
+  `  [[raffles_concerned~coat]]\n` +
+  `  Take ${n}, and no arguing about it.\n` +
+  `\n` +
+  `BUNNY:\n` +
+  `  [[bunny_realization~jacket]]\n` +
+  `  Right you are, old chap.\n` +
+  `\n`;
+
+const EXCHANGES = 30;
+const LONG_FIXTURE = Array.from({ length: EXCHANGES }, (_, i) =>
+  EXCHANGE(i + 1),
+).join("");
+
+// The whole point of the fixture: it must exceed the initial-parse viewport,
+// so a helper that reads the un-settled tree sees a truncated document every
+// single run rather than only on a loaded machine.
+const INIT_PARSE_VIEWPORT = 3000;
+
+describe("parse settling (regression guard for #281)", () => {
+  it("the fixture is longer than CodeMirror's initial-parse viewport", () => {
+    expect(LONG_FIXTURE.length).toBeGreaterThan(INIT_PARSE_VIEWPORT);
+  });
+
+  it("completeTree parses the whole document, not just the initial viewport", () => {
+    const state = EditorState.create({
+      doc: LONG_FIXTURE,
+      extensions: [language.of(SCREENPLAY_LANGUAGE_SUPPORT.language)],
+    });
+    expect(completeTree(state).length).toBe(LONG_FIXTURE.length);
+  });
+
+  it("extractPreviewText classifies every character cue, including past the initial viewport", () => {
+    const preview = extractPreviewText(LONG_FIXTURE);
+    const cues = preview
+      .split("\n")
+      .filter((l) => l.startsWith("<character>"));
+    // Two cues per exchange. A truncated tree leaves the tail unclassified
+    // (`<unknown>`), which is exactly how #281 showed up: "expected 1 to be
+    // greater than or equal to 2".
+    expect(cues.length).toBe(EXCHANGES * 2);
+    expect(preview).not.toContain("<unknown>");
+  });
+
+  // This is the invariant the DOM-reading helpers (renderPreview,
+  // extractVisibleText) and the incremental-edit tests all depend on: after
+  // settling, the tree the decoration field sees covers the whole document.
+  // Asserting it here rather than through rendered output is deliberate — a
+  // CodeMirror view only renders its viewport (under jsdom that's the first
+  // handful of lines; the rest is a `cm-gap` placeholder), so rendered output
+  // can't distinguish "parse stopped early" from "not in the viewport".
+  it("settleParse commits the complete tree into the view's state", () => {
+    const parent = document.createElement("div");
+    document.body.appendChild(parent);
+    try {
+      const view = new EditorView({
+        state: EditorState.create({
+          doc: LONG_FIXTURE,
+          extensions: [
+            language.of(SCREENPLAY_LANGUAGE_SUPPORT.language),
+            screenplayFormatting(),
+          ],
+        }),
+        parent,
+      });
+      // Sanity: without settling, the tree really is short — that's the
+      // hazard, and it's what makes this guard meaningful.
+      expect(syntaxTree(view.state).length).toBeLessThan(LONG_FIXTURE.length);
+      settleParse(view);
+      expect(syntaxTree(view.state).length).toBe(LONG_FIXTURE.length);
+      view.destroy();
+    } finally {
+      parent.remove();
+    }
+  });
+});

--- a/packages/sparkdown-document-views/test/sequential-edits.test.ts
+++ b/packages/sparkdown-document-views/test/sequential-edits.test.ts
@@ -13,7 +13,7 @@
 // one" symptom — line decorations on character cues / dialogue
 // content stop matching the lines that contain them.
 
-import { ensureSyntaxTree, language } from "@codemirror/language";
+import { language } from "@codemirror/language";
 import { EditorState } from "@codemirror/state";
 import { EditorView } from "@codemirror/view";
 import { describe, expect, it } from "vitest";
@@ -21,6 +21,7 @@ import {
   SCREENPLAY_LANGUAGE_SUPPORT,
   default as screenplayFormatting,
 } from "../src/modules/screenplay-preview/utils/screenplayFormatting";
+import { settleParse } from "./helpers/parseSettle";
 
 const mount = (source: string): EditorView => {
   const parent = document.createElement("div");
@@ -42,9 +43,9 @@ const contentHTML = (view: EditorView): string => {
   // Force the incremental parser to catch up before sampling the DOM.
   // Otherwise the partial tree leaks decorations into the trailing
   // viewport area, producing spurious classes like 17 `collapse`s on
-  // the final cm-line.
-  ensureSyntaxTree(view.state, view.state.doc.length, 30_000);
-  view.dispatch({}); // trigger a no-op update so decorate() re-runs against the full tree
+  // the final cm-line. (settleParse also commits the finished tree back into
+  // the state field so decorate() re-runs against it — see helpers/parseSettle.)
+  settleParse(view);
   const c = view.dom.querySelector(".cm-content");
   if (!c) return "";
   return c.outerHTML.replace(

--- a/packages/sparkdown-document-views/vitest.config.ts
+++ b/packages/sparkdown-document-views/vitest.config.ts
@@ -27,13 +27,25 @@ export default defineConfig({
     include: ["test/**/*.test.ts"],
     environment: "jsdom",
     pool: "threads",
-    // Run test files sequentially. These are heavy jsdom + CodeMirror tests
-    // whose rendered output depends on lezer incremental-parser warm-up state;
-    // under file parallelism the warm-up order across files is
-    // non-deterministic, which made a few separator/blank-line assertions flake
-    // intermittently. Sequential execution makes the suite deterministic. (The
-    // underlying parser warm-up sensitivity is tracked separately; this is the
-    // reliable test-harness workaround, not a timeout band-aid.)
-    fileParallelism: false,
+    // The incremental-edit tests dispatch hundreds of single-character
+    // transactions and reparse after each one; they run in a few seconds alone
+    // but comfortably exceed vitest's 5s default when several jsdom workers are
+    // competing for cores. A too-tight budget is just another way for the suite
+    // to go red without anything being broken (#281), so give every test room
+    // and let a genuine hang be the only thing that trips this.
+    testTimeout: 30_000,
+    // File parallelism is ON. It used to be disabled because these jsdom +
+    // CodeMirror tests flaked under load, which looked like parser warm-up
+    // sensitivity. The real cause (#281) was that the test helpers read
+    // CodeMirror's syntax tree straight after creating a state or view —
+    // and `@codemirror/language` budgets that initial parse by wall clock
+    // (20ms), so a busy machine yielded a truncated document and an assertion
+    // about missing content failed. Sequential execution only made the machine
+    // less busy; it narrowed the window without closing it.
+    //
+    // The helpers now parse to completion before reading (see
+    // test/helpers/parseSettle.ts, pinned by test/parse-settle.test.ts), so
+    // the result no longer depends on timing at all and parallelism is safe.
+    // It also takes the full suite from ~100s to under 20s.
   },
 });

